### DISCUSSION TITLE
Graphs with periodic boundary conditions

### DISCUSSION THREAD

**Discussion by jmmshn** [networkx/networkx#4605] [Ideas]
Hi, I'm wondering if networkx has any plans to handle graphs for systems with periodic boundary conditions in arbitrary dimensions.
The problem I'm envisioning is essentially finding valid pathways in graphs that connect across periodic boundary conditions.
This is a common problem for people working with crystal structures but I have no idea if this important for people in other fields.

**Comment by dschult** [comment]
NetworkX does have graph generators for grid graphs in n-D with periodic boundary conditions: [nx.grid_graph](https://networkx.org/documentation/stable//reference/generated/networkx.generators.lattice.grid_graph.html)

In general the "dimension" of a graph and the "boundary" can be defined in different ways. E.g. the dimension can change from one part of the graph to another. But if you are talking about mapping n-dimensional data from a continuous domain to a discrete domain represented by a network, then NetworkX can do that.

On an even more general level, you can simply connect nodes across the boundary to make edges.  The valid pathways will be found without any issues.

**Reply by jmmshn** [reply]
Thanks for the reply!
I will look at the code in `grid_graph` more closely.   My original understanding was that the setup of the grid periodic graph is on a very structured grid so the repeating cell is the same as the node which limits the usage cases.
I will look at the code more closely this weekend to see if that is the case and if it can be logically extended.

**Reply by dschult** [reply]
That's correct, we don't provide a way to create a grid of general cells -- it is too hard to design a good way to designate which nodes in each cell connect to which nodes in the neighboring cells.  We do provide reasonable tools that should make that possible for each special case though. 

Maybe....   `G1=nx.grid_graph((Nx, Ny, Nz), periodic=True)` to construct the grid, `cell = ...` for your cell/pattern.  Then loop through `G1`, constructing `G2` by adding the new cell and figuring out which nodes connect based on the neighbors in the graph G1.  Copies of the new_cell can be added with integer node names with something like:
```python
for node, nbrs in G1.adj.items():
    x, y, z = node
    N = Nx*x + Ny*y + Nz*z
    new_cell = nx.convert_node_labels_to_integers(cell, first_label=N * len(cell), label_attribute=node)
    G2.update(new_cell)
    nbrs_1st_node = [(Nx*nbrx + Ny*nbry + Nz*nbrz) for nbrx, nbry, nbrz in nbrs]
    connections = [(N+2, nbr+4) for nbr in nbrs] + [(N+5, nbr+2) for nbr in nbrs]   # this could be tricky
    G2.add_edges_from(connections)
```
The part that is hard to generalize is devising a way to state which parts of each cell connect to which parts of the neighbor cell.

**Reply by jmmshn** [reply]
So one possible way to do this is to decorate the edges with a `to_image` integer-valued vector.
Such a vector, long with a convention that the `to_image` direction always refers to the connection between the lower indexed node to the higher index node. 
There is some additional logic needed for cases when the edge is between two periodic versions of the same node but we have that figured out mostly.

Currently, the code for doing this lives at two different places in our atomic-structure specific codes and there is no clear way to organize them together in a way that people in other communities who are interested in period graphs might find useful.

https://github.com/materialsproject/pymatgen/blob/v2021.1.28/pymatgen/analysis/graphs.py
https://github.com/materialsvirtuallab/pymatgen-diffusion/blob/master/pymatgen_diffusion/neb/periodic_dijkstra.py

**Reply by dschult** [reply]
Looks like you are hosting a dijkstra algorithm (which is in networkx -- if you can get the definition of periodic pinned down) that handles periodic crystals. But I can't tell how you are imposing periodicity or even what your definition of periodic is.  Also, It looks like you have indexes as nodes and `to_image` objects to link ??image files?? with each node? I guess I'm not sure what the `to_image` attribute does for you. Why not just have the indexes as nodes and assign node and edge attributes?  But perhaps most critical to this discussion is a big-picture overview of what the nodes and edges of your graph are and how you define periodic. :}
